File: cpp/common/bypstest/TestUtils.cpp

#include "TestUtils.h"
#include "testfw.h"

using namespace com::wilutions::byps;
using namespace com::wilutions::test;

using namespace com::wilutions::byps::test::api;
using namespace com::wilutions::byps::test::api::prim;
using namespace com::wilutions::byps::test::api::remote;
using namespace com::wilutions::byps::test::api::inl;

PPrimitiveTypes internalCreateObjectPrimitiveTypes() {

    static int createCount = 1;

    PPrimitiveTypes p(new PrimitiveTypes());
    p->boolVal = (createCount & 1) != 0;
    p->intVal = ++createCount * (p->boolVal ? -1 : 1);
    p->byteVal = (int8_t)(createCount & 0xF);
    p->shortVal = (int16_t)(createCount + 100);
    p->longVal = createCount + 100000;
    p->floatVal = ((float)createCount) * 1000;
    p->doubleVal = (double)createCount * 10000;

	std::wstringstream ss;
    ss << p->intVal;
    p->stringVal = ss.str();

    return p;
}


PPrimitiveTypes TestUtils::createObjectPrimitiveTypes() {
    PPrimitiveTypes p = internalCreateObjectPrimitiveTypes();
    return p;
}

void TestUtils::releasePrimitiveTypes(com::wilutions::byps::test::api::prim::PPrimitiveTypes p) {
	p->objVal.reset();
	p->objVal2.reset();
}

int16_t TestUtils::x;

Point2D TestUtils::createPoint2D() {
	Point2D pt;
	pt.x = x++;
	pt.y = x*x;
	return pt;
}

#define TASSERT_MEMBER(name) ::tassert(file, line, msg + std::wstring(L".") + BToStdWString(#name), e->name, r->name)

void TestUtils::tassert(const char* file, int line, 
						const std::wstring& msg, 
						PPrimitiveTypes e, 
						PPrimitiveTypes r) 
{

	if ((e != NULL) ^ (r != NULL)) {
		throw AssertException(file, line, msg + L" wrong reference");
	}
	if (e == NULL) return;

	TASSERT_MEMBER(boolVal);
	TASSERT_MEMBER(byteVal);
	TASSERT_MEMBER(charVal);
	TASSERT_MEMBER(shortVal);
	TASSERT_MEMBER(intVal);
	TASSERT_MEMBER(longVal);
	TASSERT_MEMBER(floatVal);
	TASSERT_MEMBER(doubleVal);
	TASSERT_MEMBER(stringVal);

	tassert(file, line, msg + L".objVal",
		byps_static_ptr_cast<PrimitiveTypes>(e->objVal), 
		byps_static_ptr_cast<PrimitiveTypes>(r->objVal));

}

void TestUtils::tassert(const char* file, int line, 
						const std::wstring& msg, 
						const Point2D& e1, 
						const Point2D& r1)  {
	const Point2D* e = &e1;
	const Point2D* r = &r1;
	TASSERT_MEMBER(x);
	TASSERT_MEMBER(y);


}

File: java/bypstest-srv/WebContent/jasmine/spec/TestUtils.js
var createCount = 0;

var log = {
	info : function() {},
	error : function() {}
};

var TestUtils = {
	assertEquals : function(log, msg, expectedValue, readValue) {
		
	    var nnull = function(v) {
    		if (typeof v === 'object') {
    			if (v["*nnull"]) return;
    			v["*nnull"] = true;
    	    	for (ename in v) {
    	    		var e = v[ename];
    	    		var t = typeof e;
    				if (e === undefined || e === null) {
    					delete v[ename];
    				}
    				else if (t === 'boolean' && !e) {
    					delete v[ename];
    				}
    				else if (t === 'function') {
    					delete v[ename];
    				}
    				else {
    					nnull(e);
    				}
    			}
    		}
    		return v;
	    };
	    
	    var removeNullTag = function(v) {
	   		if (typeof v === 'object') {
    			if (!v["*nnull"]) return;
    			delete v["*nnull"];
       	    	for (ename in v) {
       	    		var e = v[ename];
       	    		removeNullTag(e);
       	    	}
	   		}
	   		return v;
	    };
	    
	    var trimNulls = function(v) {
	    	v = nnull(v);
	    	v = removeNullTag(v);
	    	return v;
	    };
	    
	    expectedValue = trimNulls(expectedValue);
	    readValue = trimNulls(readValue);
	    
	    // So geht's nicht, weil die Objekte rekurive Referenzen enthalten können.
	    // Rekursionen können aber nicht mit JSON serialisiert werden.
//	    var strE = JSON.stringify(expectedValue);
//	    var strR = JSON.stringify(readValue);
//	    expectedValue = JSON.parse(strE);
//	    readValue = JSON.parse(strR);
	    
		expect(expectedValue).toEqual(readValue);
	},
	
	assertTrue : function(log, msg, val) {
		if (!val) {
			expect(val).toBe(true);
		}
	},
	
	fail : function(log, msg) {
		expect(msg).toBe(null);
	},
	
	createObjectPrimitiveTypes : function() {
		var p = new byps.test.api.prim.PrimitiveTypes();
		p.intVal = ++createCount;
		p.boolVal = (createCount & 1) != 0;
		p.byteVal = (createCount * 2) & 0x7F;
		p.shortVal = (createCount * 31);
		p.longVal = "" + (createCount * 129) + ".";
		p.floatVal = createCount + 0.2;
		p.doubleVal = createCount / 7.1;
		p.stringVal = "" + p.intVal;
		p.charVal = 'abcdefghijklmnopqrstuvwxyz'.charAt(createCount%26);
		p.dateVal = new Date();
		return p;
	}
};

